src/Data/Audience.elm

```diff
@@ -34,7 +36,30 @@ type alias Audience =
     , folder : Maybe Int
     }
 
+{-| Decoder for an Audience
+-}
+decodeAudience : JD.Decoder Audience
+decodeAudience = JD.map4
+    Audience
+        (JD.field "id" JD.int)
+        (JD.field "name" JD.string)
+        (JD.field "type" decodeAudienceType)
+        (JD.field "folder" (JD.maybe JD.int))
```

Comment:
If there were more than 8 fields in an Audience to decode (so that you couldn't use `JD.map8`-like approach), what decoding approach would you choose?

Reply:
There are multiple ways how to decode a JSON object without using `mapN` in Elm. Basically, a pipeline and applicatives are a general idea here. 

```
apply : JD.Decoder a -> JD.Decoder (a -> b) -> JD.Decoder b
apply da df = JD.andThen (\f -> JD.map f da ) df

apply1 : JD.Decoder a -> JD.Decoder (a -> b) -> JD.Decoder b
apply1 da df = JD.map2 (\arg func -> func arg) da df

apply2 : JD.Decoder a -> JD.Decoder (a -> b) -> JD.Decoder b
apply2 da df = JD.map2 (|>) da df

apply3 : JD.Decoder a -> JD.Decoder (a -> b) -> JD.Decoder b
apply3 = JD.map2 (|>)

decodeAudience : JD.Decoder Audience
decodeAudience = JD.succeed Audience
    |> apply (JD.field "id" JD.int)
    |> apply1 (JD.field "name" JD.string)
    |> apply2 (JD.field "type" decodeAudienceType)
    |> apply3 (JD.field "folder" (JD.maybe JD.int))
```
JD.succeed Audience produces a value of type `Decoder a` (where, correct me if I am wrong, `a` is of type `c->d->e->f-> Audience` because Audience needs 4 parameters to be created). This value of type `Decoder a` is sent down the pipeline by operator `|>`. 
All of the apply functions are just Elm's version of `(<*>)` in Haskell (`(<*>)` is called bind for applicative in Haskell). 

The first argument for apply{ ,1,2,3} is always a (JD.Decoder a) value (for example `(JD.field "id" JD.int)`). The second argument for `apply` is a JD.Decoder (a->b) (this argument goes down the pipeline) which is `c->d->e->f-> Audience` for `apply`, `d->e->f-> Audience` for `apply1`, `e->f-> Audience` for `apply2` and `f-> Audience` for `apply3`.

Comment to `apply` function:
Function `andThen : (a -> Decoder b) -> Decoder a -> Decoder b` helps unpack the `df` function (i.e., df is of type `Decoder (c->d->e->f-> Audience)`) and so we can apply the unpacked function `f` (of type `c->d->e->f-> Audience`) to the da value (of type `Decoder a`) using simple `map : (a -> value) -> Decoder a -> Decoder value`. After applying map, the result is `Decoder (d->e->f->Audience)` which is sent to the next part in the pipeline.

Comment to `apply1` function:
Function `map2 : (a -> b -> value) -> Decoder a -> Decoder b -> Decoder value` takes as the first argument a function which takes "unpacked" values of types a and b and return some value. The second and third argument of `map2` are "packed" . `map2` unpacks those two arguments and applies a given function on the unpacked values. Then it packs the result (packing of applicatives may be done through `pure` in Haskell)

Comment to `apply2` function:
since we know, that in `apply1`, the first argument of `map2` is a function of type `a -> (a->b) -> b` we can safely use `(|>)` operator, instead of our custom function. 

Comment to `apply3` function:
We can commit the arguments since they are applied to `|>` in the same order as function `apply3` receives them.

I hope my answer was clear enough. If not, please, ask and let me explain.

I found one more neat solution: https://package.elm-lang.org/packages/webbhuset/elm-json-decode/latest/
The thing I like about this solution is that the record field order does not matter. 

Note:
My teacher (of Advanced functional programming) used terms "packed" and "unpack" to better explain, how functors, applicatives and monads work. That they are only values packed in some box and everytime we want to apply a packed function to a packed value, we need to unpack them first. Those terms helped me, therefore I use them. I hope you know what I mean.

I hope to write the rest of the comments until the end of the day. Sorry for the delay.

Reply:
Thanks for the extensive comment, yeah applicatives are the way to go.
We're using JDE.andMap pattern at work.
```elm
Json.Decode.succeed Record
 |> Json.Decode.Extra.andMap (Json.Decode.field ...)
 |> Json.Decode.Extra.andMap (Json.Decode.field ...)
 |> Json.Decode.Extra.andMap (Json.Decode.field ...)
 |> Json.Decode.Extra.andMap (Json.Decode.field ...)
 |> Json.Decode.Extra.andMap (Json.Decode.field ...)
 |> Json.Decode.Extra.andMap (Json.Decode.field ...)
 |> Json.Decode.Extra.andMap (Json.Decode.field ...)
 |> Json.Decode.Extra.andMap (Json.Decode.field ...)
 |> Json.Decode.Extra.andMap (Json.Decode.field ...)
 |> Json.Decode.Extra.andMap (Json.Decode.field ...)
```

> I found one more neat solution: https://package.elm-lang.org/packages/webbhuset/elm-json-decode/latest/
The thing I like about this solution is that the record field order does not matter.

I wish the `andThen` lambda nesting wasn't formatted so badly by `elm-format`. ie. [this ellie](https://ellie-app.com/bjC2b4nxckma1)
![](https://miro.medium.com/max/721/1*zxx4iQAG4HilOIQqDKpxJw.jpeg)